network/message_scope.go

```diff
@@ -0,0 +1,206 @@
+	return string(m)
+}
+
+const (
+	// ProtocolTypeUnicast is protocol type for unicast messages.
+	ProtocolTypeUnicast ProtocolType = "unicast"
+
+	// ProtocolTypePubSub is the protocol type for pubsub messages.
+	ProtocolTypePubSub ProtocolType = "pubsub"
+)
+
+// IncomingMessageScope captures the context around an incoming message that is received by the network layer.
+type IncomingMessageScope struct {
+	originId       flow.Identifier     // the origin node ID.
+	targetIds      flow.IdentifierList // the target node IDs (i.e., intended recipients).
+	eventId        hash.Hash           // hash of the payload and channel.
+	msg            *message.Message    // the raw message received.
+	decodedPayload interface{}         // decoded payload of the message.
+	protocol       ProtocolType        // the type of protocol used to receive the message.
+}
+
+func NewIncomingScope(originId flow.Identifier, protocol ProtocolType, msg *message.Message, decodedPayload interface{}) (*IncomingMessageScope, error) {
+	eventId, err := EventId(channels.Channel(msg.ChannelID), msg.Payload)
+	if err != nil {
+		return nil, fmt.Errorf("could not compute event id: %w", err)
+	}
+
+	targetIds, err := flow.ByteSlicesToIds(msg.TargetIDs)
+	if err != nil {
+		return nil, fmt.Errorf("could not convert target ids: %w", err)
+	}
+	return &IncomingMessageScope{
+		eventId:        eventId,
+		originId:       originId,
+		msg:            msg,
+		decodedPayload: decodedPayload,
+		protocol:       protocol,
+		targetIds:      targetIds,
+	}, nil
+}
+
+func (m IncomingMessageScope) OriginId() flow.Identifier {
+	return m.originId
+}
+
+func (m IncomingMessageScope) Proto() *message.Message {
+	return m.msg
+}
+
+func (m IncomingMessageScope) DecodedPayload() interface{} {
+	return m.decodedPayload
+}
+
+func (m IncomingMessageScope) Protocol() ProtocolType {
+	return m.protocol
+}
+
+func (m IncomingMessageScope) ChannelId() channels.Channel {
+	return channels.Channel(m.msg.ChannelID)
+}
+
+func (m IncomingMessageScope) Size() int {
+	return m.msg.Size()
+}
+
+func (m IncomingMessageScope) TargetIDs() flow.IdentifierList {
+	return m.targetIds
+}
+
+func (m IncomingMessageScope) EventID() []byte {
+	return m.eventId[:]
+}
+
+func (m IncomingMessageScope) PayloadType() string {
+	return MessageType(m.msg.Payload)
+}
+
+// OutgoingMessageScope captures the context around an outgoing message that is about to be sent.
+type OutgoingMessageScope struct {
+	targetIds flow.IdentifierList               // the target node IDs.
+	channelId channels.Channel                  // the channel ID.
+	payload   interface{}                       // the payload to be sent.
+	encoder   func(interface{}) ([]byte, error) // the encoder to encode the payload.
+	msg       *message.Message                  // raw proto message sent on wire.
+	protocol  ProtocolType                      // the type of protocol used to send the message.
+}
+
+func NewOutgoingScope(
```

Comment:
please document the return errors

Reply:
https://github.com/onflow/flow-go/pull/3675/commits/f18504204c7df509d213cdc1e67cacf670161e04